crates/metrics/src/label.rs

```diff
@@ -0,0 +1,545 @@
+    pub fn resolve_label<T>(&self, label: T) -> &M
+    where
+        Self: ResolveLabels<(T,), Target = M>,
+    {
+        self.resolve_labels((label,))
+    }
+
+    /// Resolves multilpe dynamic labels and finds the underlying metric.
+    ///
+    /// Expects a tuple of labels as `LS` argument.
+    pub fn resolve_labels<LS>(&self, labels: LS) -> &<Self as ResolveLabels<LS>>::Target
+    where
+        Self: ResolveLabels<LS>,
+    {
+        ResolveLabels::resolve_labels(self, labels)
+    }
+}
+
+/// Metric label using an `enum` as its value.
+///
+/// The most efficient way to specify metric labels in runtime, metric lookups
+/// using enum labels are as fast as array indexing. Prefer using this when all
+/// of your possible label values are known at the compile time.
+#[derive(Clone, Copy, Debug)]
+pub struct EnumLabel<const NAME: LabelName, T>(T);
+
+impl<const NAME: LabelName, T> EnumLabel<NAME, T> {
+    /// Creates a new [`EnumLabel`].
+    pub fn new(e: impl Into<T>) -> Self {
+        Self(e.into())
+    }
+
+    /// Convert this [`EnumLabel`] into the inner [`Enum`].
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+/// `enum` used as the value in [`EnumLabel`]s.
+///
+/// To implement this `trait` you also need to derive [`Ordinalize`] for your
+/// `enum`.
+/// SAFETY: DO NOT use custom discriminant values (eg. `enum MyEnum { MyVariant
+/// = -1 }`), this will lead to either:
+/// - `panic` in runtime (only for builds with `debug_assertions`)
+/// - incorrect label resolution
+pub trait Enum: Copy + Ordinalize<VariantType = i8> {
+    /// String representation of this enum.
+    fn as_str(&self) -> &'static str;
+}
+
+impl<const NAME: LabelName, T, M> DynamicLabel<M> for EnumLabel<NAME, T>
+where
+    T: Enum,
+{
+    // TODO: Switch to `[(T, M); T::VARIANT_COUNT]` once generic parameters are
+    // allowed to be used in const expressions.
+    type MetricCollection = Vec<(T, M)>;
+}
+
+impl<const NAME: LabelName, T, M> Metric for WithLabel<EnumLabel<NAME, T>, M>
+where
+    T: Enum,
+    M: Metric,
+{
+    fn register(attrs: &Attrs) -> Self {
+        let name = const { resolve_label_name::<NAME>() };
+
+        let metrics = T::VARIANTS.iter().map(|l| {
+            let label = Label::from_static_parts(name, l.as_str());
+            (*l, M::register(&attrs.with_label(label)))
+        });
+
+        Self {
+            collection: metrics.collect(),
+        }
+    }
+}
+
+impl<const NAME: LabelName, T, M> ResolveLabels<(EnumLabel<NAME, T>,)>
+    for WithLabel<EnumLabel<NAME, T>, M>
+where
+    T: Enum,
+    M: Metric,
+{
+    type Target = M;
+
+    fn resolve_labels(&self, (label,): (EnumLabel<NAME, T>,)) -> &M {
+        let debug_panic = || {
+            if cfg!(debug_assertions) {
+                panic!("Invalid enum usage, custom discriminants must not be used")
+            }
+        };
+
+        let idx = label.0.ordinal();
+        let mut idx = if idx < 0 {
```

Comment:
Why not use `debug_assert!(idx >= 0)`?

Reply:
Because I'm also doing `idx = 0` in the brach, so the release build won't crash

Reply:
That can be replaced by `idx.max(0)`.

Reply:
We need to check that the index is in `[0, vec.len())` range